src/python/pants/engine/internals/parser.py

```diff
@@ -152,26 +158,77 @@ def parse(
                 f"{original}.\n\n{help_str}\n\nAll registered symbols: {valid_symbols}"
             )
 
-        error_on_imports(build_file_content, filepath)
-
         return self._parse_state.parsed_targets()
 
 
-def error_on_imports(build_file_content: str, filepath: str) -> None:
-    # This is poor sandboxing; there are many ways to get around this. But it's sufficient to tell
-    # users who aren't malicious that they're doing something wrong, and it has a low performance
-    # overhead.
-    if "import" not in build_file_content:
-        return
-    io_wrapped_python = StringIO(build_file_content)
-    for token in tokenize.generate_tokens(io_wrapped_python.readline):
-        token_str = token[1]
-        lineno, _ = token[2]
-        if token_str != "import":
-            continue
+def error_on_imports_and_rewrite_caofs_to_avoid_ambiguous_symbols(
+    content: str, filepath: str, *, rename_symbols: bool
+) -> str:
+    """Error if imports are detected, which break sandboxing, and rewrite the symbol of context-
+    aware object factories to avoid ambiguity with target generators.
+
+    We use a simple heuristic of checking if `name` is set. If not, we assume it's a macro.
+    """
+    tree = ast.parse(content, filename=filepath)
+    visitor = AstVisitor(filepath, rename_symbols=rename_symbols)
+    visitor.visit(tree)
+    if not visitor.pending_renames:
+        return content
+    lines = content.splitlines(keepends=True)
+    for rename in visitor.pending_renames:
+        idx = rename.lineno - 1
+        lines[idx] = lines[idx].replace(f"{rename.old_symbol}(", f"{rename.new_symbol}(")
+    return "".join(lines)
+
+
+DEPRECATED_MACRO_RENAMES = {
+    "python_requirements": "python_requirements_deprecated_macro",
+    "poetry_requirements": "poetry_requirements_deprecated_macro",
+    "pipenv_requirements": "pipenv_requirements_deprecated_macro",
+    "pants_requirement": "pants_requirement_deprecated_macro",
+}
```

Comment:
This is not very generic, but it captures all the CAOFs in use without adding import cycles or having to dynamically query. Also this code is temporary.

> having to dynamically query

We don't want to do this rewrite for user plugins. The scheme would almost certainly fail.